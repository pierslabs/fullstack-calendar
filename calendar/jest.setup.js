import 'whatwg-fetch';

import 'setimmediate';

require('dotenv').config({
  path: 'test.env',
});

// Realizar el mock completo de las variables de entorno
jest.mock('./src/helpers/getEnvVar', () => ({
  getEnvVar: () => ({ ...process.env }),
}));
