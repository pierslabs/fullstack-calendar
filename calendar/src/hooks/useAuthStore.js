import { useSelector, useDispatch } from 'react-redux';
import calendarAPi from '../api/calendarApi';
import {
  clearErrorMessage,
  onChecking,
  onLogin,
  onLogOutCalendar,
  onLogut,
} from '../store';

//Alternative method thunks

export const useAuthStore = () => {
  const { status, user, errorMessage } = useSelector((state) => state.auth);

  const dispatch = useDispatch();

  const startLogin = async ({ email, password }) => {
    dispatch(onChecking());

    try {
      const { data } = await calendarAPi.post('/auth', { email, password });

      localStorage.setItem('token', data.token);
      localStorage.setItem('token-init-date', new Date().getTime());

      dispatch(onLogin({ name: data.name, uid: data.uid }));
    } catch (error) {
      dispatch(onLogut(error.response.data?.msg || '--'));
      setTimeout(() => {
        dispatch(clearErrorMessage());
      }, 10);
    }
  };

  const startRegister = async ({ email, password, name }) => {
    dispatch(onChecking());
    try {
      const { data } = await calendarAPi.post('/auth/new', {
        email,
        password,
        name,
      });

      localStorage.setItem('token', data.token);
      localStorage.setItem('token-init-date', new Date().getTime());

      dispatch(onLogin({ name: data.name, uid: data.uid }));
    } catch (error) {
      dispatch(onLogut(error.response.data?.msg || '---'));
      setTimeout(() => {
        clearErrorMessage();
      }, 10);
    }
  };

  const checkAuthToken = async () => {
    const token = localStorage.getItem('token');
    if (!token) return dispatch(onLogut());

    try {
      const { data } = await calendarAPi.get('/auth/renew');

      localStorage.setItem('token', data.token);
      localStorage.setItem('token-init-date', new Date().getTime());
      dispatch(onLogin({ name: data.name, uid: data.uid }));
    } catch (error) {
      dispatch(onLogut(error.response.data?.msg || '---'));
      setTimeout(() => {
        clearErrorMessage();
      }, 10);
    }
  };

  const startLogOut = () => {
    localStorage.clear();
    dispatch(onLogOutCalendar());
    dispatch(onLogut());
  };

  return {
    status,
    user,
    errorMessage,

    //methods
    startLogin,
    startRegister,
    checkAuthToken,
    startLogOut,
  };
};
