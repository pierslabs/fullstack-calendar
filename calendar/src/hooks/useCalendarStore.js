import { useDispatch, useSelector } from 'react-redux';
import Swal from 'sweetalert2';
import calendarAPi from '../api/calendarApi';
import { convertDateEvents } from '../helpers';
import {
  onAddNewEvent,
  onDeleteEvent,
  onLoadEvents,
  onSetActiveEvent,
  onUpdateEvent,
} from '../store/calendar/calendarSlice';

export const useCalendarStore = () => {
  const dispatch = useDispatch();

  const { events, activeEvent } = useSelector((state) => state.calendar);
  const { user } = useSelector((state) => state.auth);

  const setActiveEvent = (calendarEvent) => {
    dispatch(onSetActiveEvent(calendarEvent));
  };

  const startSavingEvent = async (calendarEvent) => {
    try {
      if (calendarEvent.id) {
        await calendarAPi.put(`/events/${calendarEvent.id}`, {
          ...calendarEvent,
        });
        dispatch(onUpdateEvent({ ...calendarEvent, user }));
        return;
      }
      //create
      const { data } = await calendarAPi.post('/events', calendarEvent);
      dispatch(onAddNewEvent({ ...calendarEvent, id: data.evento.id, user }));
    } catch (error) {
      Swal.fire('Error save event', error.response.data?.msg, 'error');
    }
  };

  const startLoadingEvents = async () => {
    try {
      const { data } = await calendarAPi.get('/events');
      const events = convertDateEvents(data.eventos);
      dispatch(onLoadEvents(events));
    } catch (error) {}
  };

  const startDeleteEvent = async () => {
    try {
      await calendarAPi.delete(`/events/${activeEvent.id}`);
      dispatch(onDeleteEvent());
    } catch (error) {
      Swal.fire('Error delete event', error.response.data?.msg, error);
    }
  };

  return {
    //*Properties
    events,
    activeEvent,
    hasEventSelected: !!activeEvent?.id,
    //*Methods
    setActiveEvent,
    startSavingEvent,
    startDeleteEvent,
    startLoadingEvents,
  };
};
