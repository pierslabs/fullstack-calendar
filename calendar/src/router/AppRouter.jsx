import React, { useEffect } from 'react';
import { Navigate, Route, Routes } from 'react-router-dom';
import LoginPage from '../auth/pages/LoginPage/LoginPage';
import { CalendarPage } from '../calendar/pages';
import { useAuthStore } from '../hooks';

const AppRouter = () => {
  const { checkAuthToken, status } = useAuthStore();

  useEffect(() => {
    checkAuthToken();
  }, []);

  if (status === 'checking') {
    return <h1>Cargando...</h1>;
  }
  return (
    <Routes>
      {status === 'not-authenticated' ? (
        // when token doesn't exist route not exists too
        <>
          <Route path="/auth/*" element={<LoginPage />} />
          <Route path="/*" element={<Navigate to="/auth/login" />} />
        </>
      ) : (
        <>
          <Route path="/" element={<CalendarPage />} />
          <Route path="/*" element={<Navigate to="/" />} />
        </>
      )}
    </Routes>
  );
};

export default AppRouter;
