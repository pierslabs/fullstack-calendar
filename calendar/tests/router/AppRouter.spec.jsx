import { render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { CalendarPage } from '../../src/calendar/pages';
import { useAuthStore } from '../../src/hooks/useAuthStore';
import AppRouter from '../../src/router/AppRouter';

jest.mock('../../src/hooks/useAuthStore');

jest.mock('../../src/calendar/pages/CalendarPage', () => ({
  CalendarPage: () => <h1>CalendarPage</h1>,
}));

describe('AppRputer', () => {
  const mockCheckAuthToken = jest.fn();

  beforeEach(() => jest.clearAllMocks());

  test('should return default routes', () => {
    /*
    importante cuando hay plugins de terceros a veces que se establecen en el elemnto root
    como por ejemplo el modal,al final es ignorarlo si estamos en testing porque realmente
    el equipo de desarrollo del plugin  ya tendrian que haber realizado ellos las pruebas pertinentes
    otra solucion es mockear todo el plugin, ir a calerndar Modal para ver solucion implementada
    */

    useAuthStore.mockReturnValue({
      status: 'checking',
      checkAuthToken: mockCheckAuthToken,
    });
    render(<AppRouter />);

    expect(screen.getByText('Cargando...')).toBeTruthy();
  });

  test('should return Login if not autenticated', () => {
    /**
     * <MemoryRouter initialEntries={['/auth/']}>
     *  <AppRouter />
     * </MemoryRouter>
     */
    useAuthStore.mockReturnValue({
      status: 'not-authenticated',
      checkAuthToken: mockCheckAuthToken,
    });
    const { container } = render(
      <MemoryRouter initialEntries={['/auth/']}>
        <AppRouter />
      </MemoryRouter>
    );
    expect(screen.getByText('Ingreso')).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  test('should return Calendar if autenticated', () => {
    /**
     * <MemoryRouter initialEntries={['/auth/']}>
     *  <AppRouter />
     * </MemoryRouter>
     */
    useAuthStore.mockReturnValue({
      status: 'authenticated',
      checkAuthToken: mockCheckAuthToken,
    });
    render(
      <MemoryRouter initialEntries={['/auth/']}>
        <AppRouter />
      </MemoryRouter>
    );

    /**
     * Nos dara un error en Calendar si no mockeamos todos los hooks
     * como deberian estar ya probados y seria muy tedioso hacerlos todos
     * haremos un mock directamente del calendar page
     */
    // screen.debug();

    expect(screen.getByText('CalendarPage')).toBeTruthy();
  });
});
