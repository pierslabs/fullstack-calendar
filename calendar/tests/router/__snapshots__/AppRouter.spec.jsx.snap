// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppRputer should return Calendar if autenticated 1`] = `
<div>
  <div
    class="container login-container"
  >
    <div
      class="row"
    >
      <div
        class="col-md-6 login-form-1"
      >
        <h3>
          Ingreso
        </h3>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="loginEmail"
              placeholder="Email"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="loginPassword"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <div
            class="d-grid gap-2"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Login"
            />
          </div>
        </form>
      </div>
      <div
        class="col-md-6 login-form-2"
      >
        <h3>
          Registro
        </h3>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerName"
              placeholder="Name"
              required=""
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerEmail"
              placeholder="Email"
              required=""
              type="email"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerPassword"
              placeholder="password"
              required=""
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerPassword2"
              placeholder="Repita la contraseña"
              required=""
              type="password"
              value=""
            />
          </div>
          <div
            class="d-grid gap-2"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Crear cuenta"
            />
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;

exports[`AppRputer should return Login if not autenticated 1`] = `
<div>
  <div
    class="container login-container"
  >
    <div
      class="row"
    >
      <div
        class="col-md-6 login-form-1"
      >
        <h3>
          Ingreso
        </h3>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="loginEmail"
              placeholder="Email"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="loginPassword"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <div
            class="d-grid gap-2"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Login"
            />
          </div>
        </form>
      </div>
      <div
        class="col-md-6 login-form-2"
      >
        <h3>
          Registro
        </h3>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerName"
              placeholder="Name"
              required=""
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerEmail"
              placeholder="Email"
              required=""
              type="email"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerPassword"
              placeholder="password"
              required=""
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerPassword2"
              placeholder="Repita la contraseña"
              required=""
              type="password"
              value=""
            />
          </div>
          <div
            class="d-grid gap-2"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Crear cuenta"
            />
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
